### ISSUE TITLE
Problem with imgsz and rect=True

### ISSUE THREAD

**Issue by YEONCHEOL-HA** [ultralytics/ultralytics#9022]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I know that YOLOv8's default image size is 640*640. 

However, I saw in the YOLOv8 doc that training is possible with a rectangular image if i set rect=true. 

1. can i use rectangular image in YOLOv8 train?
2. I just wonder how images are processed in this process(rect=true).
3.  If rect=true is set, will padding work on the empty space when the original image is resized to 630*320? 

and then, If padding works, what type padding is used?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hello! 😊 Let's address your questions about using rectangular images with YOLOv8.

1. **Rectangular Images for Training**: Yes, you can indeed use rectangular images for training by setting `rect=True`. This allows for training on non-square images effectively.

2. **Image Processing with `rect=True`**: When `rect=True` is set, images are resized to match the longer dimension set by `imgsz`, maintaining their aspect ratio. This means the shorter sides might not be exactly to `imgsz`, but will be scaled accordingly.

3. **Padding and `rect=True`**: If there's empty space after resizing due to aspect ratio differences, padding is applied to maintain uniformity in dimensionality across the dataset for batch processing. YOLO uses zero-padding (pads with zeros) to fill these areas without affecting the image content.

Here's a quick code snippet for setting `rect` and `imgsz` for rectangular training:

```python
# Important: Ensure to have the latest version of Ultralytics YOLOv8
from ultralytics import YOLO

# Load your model and set imgsz (for example, 630x320) and rect=True for training
model = YOLO('path/to/your/model.yaml')
model.train(data='your_dataset.yaml', epochs=100, imgsz=(630, 320), rect=True)
```

Feel free to experiment and observe how it impacts your model's performance. Happy training!

**Comment by YEONCHEOL-HA** [comment]
thank you for your answer

i have one more question about rect option.

My question is, if I have images of various sizes, if I set imgsz to 640 and rect=true, will the long side be set to 640 and maintain the ratio?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA yes, that's correct! When you set `imgsz=640` and `rect=True`, YOLOv8 will resize your images so that the longer side of each image is scaled to 640 pixels, while maintaining the original aspect ratio for all images. This ensures that images of various sizes are appropriately scaled for the training process without distorting their aspect ratios. Happy training! 😊

**Comment by YEONCHEOL-HA** [comment]
I know that when labeling a rectangular image, each side is normalized from 0 to 1. as mentioned earlier, how does the normalized coordinate system change when zero padding is converted to 640*640?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hello there! 🌟

1. When we apply zero-padding to make rectangular images square (640x640), the normalization remains unchanged. The original content's normalized coordinates stay the same, as padding only adds empty space around the image without affecting its content.

2. If you're looking to train a model from scratch without pre-trained weights, simply specify `pretrained=False` when loading your model. Here's a quick example:

   ```python
   from ultralytics import YOLO
   
   # Load a model architecture without pre-trained weights
   model = YOLO('path/to/model.yaml', pretrained=False)
   ```

Hope this helps! If you have any more questions, feel free to ask. Happy coding!

**Comment by YEONCHEOL-HA** [comment]
and then in question 1, if the rectangular image is normalized from 0 to 1, the image isn't square due to original ratio(width and height)?

When I normalize the length and height of a rectangular image from 0 to 1, I am confused as to whether it changes at the same rate as a square.

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! 😊 Let's dive into your questions:

Q1. Yes, normalization from 0 to 1 is indeed relative, keeping the aspect ratio intact. Both width and height are normalized independently to keep their original relationship. Think of it as scaling the image down so that its longest side fits within the 0 to 1 interval, without altering the image's shape.

Q2. If `rect=false`, images are resized to square (both sides equal to `imgsz`), often with padding added to maintain aspect ratios.

Q3. & Q4. The difference between model configurations (n, s, m, l, x) lies in their size and complexity. Larger models (e.g., 'l', 'x') have more layers/parameters, aiming for higher accuracy but requiring more computational resources. Smaller ones (e.g., 'n', 's') are lighter and faster, suitable for less powerful devices or real-time applications. They have different architectures tailored to balance between speed and accuracy.

Q5. For a large dataset unrelated to COCO's pretrained classes, it can be beneficial to train from scratch, especially if the classes are very different. However, starting with pretrained weights might still offer faster convergence and potentially better initial feature detectors. Experimenting with both approaches could help you find what works best for your dataset.

Your curiosity is awesome, and I'm here to help you navigate through it! Good luck with your project! 🚀

**Comment by YEONCHEOL-HA** [comment]
thanks for your answer. 

I have few more questions! 

Q1. When receiving a 640*640 image as input in yolov8, I am curious as to whether the process is to convolution the entire image or only the image within the bounding box.

Q2. I am curious as to whether only the inside of the bbox is trained or the entire image is trained.

Q3. I'm also curious about when the bbox is used if training the entire image.

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! 😊 No worries, I'm glad to help out with your questions!

Q1 & Q2: In YOLOv8, the convolution operations are applied to the entire image, not just within the bounding boxes (bboxes). The whole image is utilized to learn and extract features that can help in identifying objects anywhere on it.

Q3 & Q4: Bounding boxes come into play during the loss calculation phase. They're used as ground truth to compare against the predicted boxes. The model learns to adjust its predictions to match these ground truths more closely, thereby improving over time.

Hope that clears things up a bit. Cheers! 🚀

**Comment by YEONCHEOL-HA** [comment]
I got the following answer from the question about normalization of rectangular image.

"Yes, normalization from 0 to 1 is indeed relative, keeping the aspect ratio intact. Both width and height are normalized independently to keep their original relationship. Think of it as scaling the image down so that its longest side fits within the 0 to 1 interval, without altering the image's shape."

I just wonder if the following image matches what it means with the answer. follwing image is about normalization of rectangular image.

![image](https://github.com/ultralytics/ultralytics/assets/63231666/3953432d-f5f4-4e09-9e0e-441a6c12c069)

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hi there! 😊

Absolutely, your understanding is on point! When we normalize rectangular images, each side (width and height) gets scaled down independently, ensuring the aspect ratio remains unchanged. So, yes, the image you've referred to aligns with the explanation given. This way, no matter the original dimensions, they fit within a 0 to 1 range, preserving their shape after normalization. If there are any more questions or anything else I can help with, feel free to reach out! Happy coding! 🚀

**Comment by YEONCHEOL-HA** [comment]
Thanks for your answer!

I got few more question about head of YOLOv8. 

Q1. In the Head, Between detection and classification, which is performed earlier?

Q2. Is it correct that ,In the def forward code, the cv2 module generates predicted values for bounding boxes, and the cv3 module generates predicted values for classes?

Q3. Is it correct that object detection and class classification occur independently, but not simultaneously?

Q4. Is the labeling file (txt) used in the def decode_bboxes code during train or prediction?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! 🌟 I'm glad you're diving deep into YOLOv8's details! Let's tackle your questions:

Q1. In the YOLOv8 head, detection (locating objects and bounding boxes) and classification (identifying object classes) are performed together in a single forward pass for efficiency.

Q2. Actually, in YOLOv8's `forward` code, we don't use `cv2` or `cv3` modules directly for predictions. YOLOv8 has its own mechanism for generating predicted bounding boxes and classes based on the model's architecture and trained weights.

Q3. Object detection (bounding boxes) and classification (class scores) happen in parallel within the network's architecture. They are integrated but performed through different parts of the network's head.

Q4. The labeling file (`.txt`) is used both during training (to provide ground truth for learning) and prediction (to decode and interpret bounding boxes in evaluation).

Q5. YOLOv8 outputs feature maps at three different scales (80x80, 40x40, and 20x20) to effectively detect objects of various sizes. Smaller objects can be recognized in denser (e.g., 80x80) grids, while larger ones are better detected in coarser (e.g., 20x20) grids.

Q6. Yes, Non-Maximum Suppression (NMS) is used after the detection process to filter out overlapping bounding boxes, keeping only the most probable one for each detected object. It's performed post-detection to refine the final object detection results.

Hope this helps clarify things for you! If you have more questions or need further explanation, feel free to ask. Happy coding! 🚀

**Comment by YEONCHEOL-HA** [comment]
Honestly, I don't understood that detection and classification are performed together. 

Q1. I thought the detection of objects was preceded and the classification of objects inside the bounding box was preformed. 

Q2. How is it possible to classify an object when you don't know its location in the entire image?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! Let's clarify your questions 🌞

1. **Detection and Classification Together**: YOLOv8 actually performs detection (locating objects) and classification (identifying what the objects are) simultaneously. This is achieved through a single neural network that outputs both the object's location (as bounding boxes) and its class. This integrated approach is part of what makes YOLO fast and efficient.

2. **Classification Without Prior Coordinates**: The YOLO architecture is designed to predict classes and bounding boxes in parallel. During training, it learns to recognize patterns associated with different object classes directly from the raw pixel values across the entire image. For an insightful dive into how YOLO performs object detection and classification, you might find [YOLO's original paper](https://arxiv.org/abs/1506.02640) a great read!

3. **Impact of Image Resizing on Learning**: When images are resized (either down or up) to the input size (`imgsz`), there could indeed be a loss or distortion of information, depending on how drastic the size change is. This might affect learning, particularly for small objects that might become harder to recognize. The diversity in your training data, including a wide range of object sizes and shapes, can help mitigate this. If your dataset has a bias towards small or large bounding boxes, it could influence the model's ability to generalize well across different sizes. Ensuring your training data is as varied and representative as possible of the real-world scenarios you expect the model to handle is key.

For example, if most bboxes in your training set are small and you're looking to detect larger objects (or vice versa), consider augmenting your dataset with more varied examples to cover the spectrum of object sizes you aim to detect.

I hope this helps clarify things! Dive into experimentation and keep exploring; that's the best way to uncover the full potential of YOLOv8. Happy detecting! 🚀

**Comment by YEONCHEOL-HA** [comment]
![YOLOv8 archtecture](https://github.com/ultralytics/ultralytics/assets/63231666/bcb4d325-132c-444f-b8c6-5ee15d5640d2)

The following image is a summary of everything you have explained to me so far.

Q1. Is the NMS appropriately located?

Q2. Is the head summary appropriate?


thank you  for your answer!

**Comment by glenn-jocher** [comment]
Hey @YEONCHEOL-HA! 👋 Thanks for reaching out.

Absolutely! The diagram you've shared superbly encapsulates the YOLOv8 architecture. 🌟

Q1. Yes, the location for NMS (Non-Maximum Suppression) is spot on. NMS is applied after detection predictions are made to refine the bounding boxes by eliminating overlaps, ensuring each detected object gets only the best possible bounding box.

Q2. The head summary looks accurate as well. It nicely outlines how the model concurrently processes detection and classification tasks, aligning well with YOLOv8's integrated approach.

If any specific questions or if a deeper dive into any part of the process is needed, I'm here to help! Keep up the great work exploring YOLOv8. Happy coding! 🚀

**Comment by YEONCHEOL-HA** [comment]
![Head summary](https://github.com/ultralytics/ultralytics/assets/63231666/4736af8e-3fe8-48d6-893d-f170d5b9ea88)

thanks for your answer!! 

I got few more question about YOLOv8. figure is head summary for YOLO detection. I almost understood about simultaneous cal culation between classification and detection. 

Q1. I read original YOLO paper. after Pr(Classi|Object) is calculated in each grid cell, How each probability is integrated as final class score? 

Q2. Is the integrated final class score(Conditional Class Probabilities) assigned based on the gird cell with the center of the bbox?

**Comment by YEONCHEOL-HA** [comment]
thank you for answer! 

Q1. when the class probability map is calculated,is the final probability  calculated by dividing the image into 3*3 grids and averaging the probability for each grid cell with object? If not, recognize that 4 boxes belong to the same class and calculate the probability that 4 boxes belong to specific class?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! 🌟

Glad you reached out with your question! In YOLOv8, the class probability map is calculated for each grid cell individually. Each cell predicts the probabilities for all classes independently. The model does not average probabilities across 3x3 grids or assume multiple cells belong to the same class for probability calculations.

For cells with detections (like your dinosaur example), YOLOv8 uses the confidence score (objectness * IOU) combined with the class probability (`Pr(Classi|Object)`) to determine the class of each detection. The detection with the highest combined score in overlapping areas (after applying Non-Maximum Suppression, NMS) is typically chosen, ensuring that each detected object gets the most accurate class assignment and bounding box.

So, for your dinosaur detected in 4 grid cells, it's the individual scores that matter, and NMS will help in picking the most accurate detection among those cells. No averaging or grouping of grid cells is done for class probability calculations. Hope this clarifies things! 😀

If you have more questions or need further examples, feel free to ask!

**Comment by YEONCHEOL-HA** [comment]
thank for your answer!
![image](https://github.com/ultralytics/ultralytics/assets/63231666/bb4635c0-c3dd-4a95-8eab-e7aac261f506)

I got few more question about YOLOv8 detection. 

Q1. you said that "he detection with the highest combined score in overlapping areas (after applying Non-Maximum Suppression, NMS) is typically chosen, ensuring that each detected object gets the most accurate class assignment and bounding box.". then in this figure, 4 grid cells is belonging to dinosaur. When I run classification on the class probabilities map, is the probability given if the score of the box (3row 3 column) at the bottom right is the highest?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! 😄 It's fantastic you're diving deep into the YOLOv8 details!

Yes, you've got it right! In the scenario you described, if the grid cell at the bottom right (3rd row, 3rd column) has the highest score for the dinosaur after NMS is applied, that’s the detection that will be chosen. Non-Maximum Suppression ensures that among overlapping detections for the same class, only the one with the highest score (combination of confidence and class probability) is kept. This process significantly enhances the accuracy of the bounding box and class assignment for each detected object.🦖

Hope this clears things up! Should you have any more queries or need further assistance, feel free to ask. Happy to help!

**Comment by YEONCHEOL-HA** [comment]
thanks for your answer!

Q1. Then I know that YOLOv1 does anchor free detection. Does YOLOv8 also follow the same principles as the previous question?

**Comment by YEONCHEOL-HA** [comment]
Q1. Are anchor boxes used during training in YOLOv8, but not relied upon during detection?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! 😊 Yes, YOLOv8 utilizes anchor boxes during the training phase to help the model learn general shapes and sizes of objects. However, the reliance on anchors is reduced during prediction, thanks to the robust feature extraction and direct bounding box regression capabilities of the model. This approach allows YOLOv8 to maintain high precision while being efficient in detecting objects of various shapes and sizes. Hope this answers your question!

**Comment by YEONCHEOL-HA** [comment]
thanks for your answer!

i got few more question about YOLOv8.

Q1. In detection model, I know that YOLOv8 employs classfication and localization simultaneously. But i confused about classification in detection. During classification, YOLOv8 still use grid that dividing image?

**Comment by glenn-jocher** [comment]
Hey there! 😊 Great questions!

Q1. Yes, during detection, YOLOv8 continues to use a grid system where each cell in the grid is responsible for predicting objects whose center falls within its bounds. This assists in localization and classification simultaneously.

Q2. In YOLOv8, classification within each grid cell is executed by predicting class probabilities for each predetermined anchor box based on the features extracted by the deep neural network. The model assesses the likelihood of each object class being present in the cell, which is integrated with the bounding box regression to determine both the presence and class of objects within each cell.

Feel free to drop more questions if you have any! Happy detecting! 🚀

**Comment by YEONCHEOL-HA** [comment]
Thanks for your answer!!


If an object exists beyond the cell, how is the classification performed? Does the algorithm only use the probability of classification for the cell where the center is located?

**Comment by glenn-jocher** [comment]
Hey there! 😊 You're welcome!

In YOLOv8, the classification is primarily focused on the cell where the object's center is located. Each cell predicts the probabilities and other attributes (like bounding boxes) for objects based on their centers falling within that specific cell. The detection and classification are tied to these center-based grid cells to help accurately localize and identify objects.

If you have more questions or need further clarification, feel free to ask. Happy to help!

**Comment by YEONCHEOL-HA** [comment]
So, does this mean that when there is an object that occupies two or more cells, classification is performed using only the features in the cell with the center in the feature map? 

Or does the algorithm use multiple cells of the feature map in classifiaction?

**Comment by glenn-jocher** [comment]
Hey there! 😊

Yes, you've got it right! In YOLOv8, for each object, the classification and bounding box predictions are primarily focused on the cell where the object's center falls in the feature map. While each cell predicts independently, the model's architecture ensures spatial coherence, so even if an object spans multiple cells, the classification remains robust due to the connected nature of convolutional layers. This approach helps in efficient detection without having to aggregate information across cells explicitly.

Hope this clears things up! Feel free to reach out if you have more questions. Happy to help! 🚀

**Comment by YEONCHEOL-HA** [comment]
Q1.  In Train, how is Bbox used for Classification? Does the algorithm learn the feature map in the bbox?

**Comment by glenn-jocher** [comment]
@YEONCHEOL-HA hey there! 😊

Q1. In YOLOv8, during training, the bounding box (bbox) primarily assists in predicting the object's location and size, which is tightly coupled with classification. The algorithm indeed learns to extract features from the area within the bbox, which includes classifying what object the bbox contains.

Q2. Yes, that's a good understanding! Even if an object spans across multiple cells, YOLOv8 can effectively classify it due to the connected nature of convolutional layers and the coherent extraction of features across the feature map. Additionally, the use of image pyramids allows the model to detect objects at various scales, enhancing detection across different object sizes and spatial distributions.

Hope this helps clarify your questions! If you have more, feel free to ask. 🌟

**Comment by YEONCHEOL-HA** [comment]
thanks you for your answer!

Q1. I just wonder in detection model. can algorism only do detection not classification?

**Comment by glenn-jocher** [comment]
Hey there! 😊

Q1. In the YOLOv8 detection model, localization and classification are typically performed together as part of the object detection process. However, if you only need localization, you could modify the model to ignore class predictions, essentially treating every detection as a single class.

Q2. The main difference is that a classification model predicts what class an entire image belongs to, while classification in a detection model predicts the classes of multiple objects within different regions of the same image.

Q3. Yes, the YOLOv8 classification model can perform multi-label classification, allowing it to predict multiple labels for a single image, which is useful for complex scenes with multiple applicable categories.

Hope this helps! Let me know if you have more questions.

**Comment by AlexShaGH** [comment]
Greetings. I’m trying to understand what the best combination of training and inference parameters pertaining to image size for my specific imagery is. The training dataset is 16 bits depth images of the identical static size – 720 x 1280. The images coming for inference will be of the same specification. The model will be quantized to FP16 and deployed on edge device. It is my understanding that for the best speed, it’s better to specify the exact image size for training and for inference. How do I do this? If I issue
model.train(…, imgsz=(720, 1280), rect=True) or model.train(…, imgsz=[720, 1280], rect=True), I get a warning:

WARNING ⚠️ updating to 'imgsz=1280'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'`

There is also a warning that it cannot be done on multiple GPUs and will default to rect=False. This I can solve by specifying only one cuda device.

Please, advise on how exactly to set the parameters for training and for inference so the model performs the best on static size imagery on edge device.

**Comment by glenn-jocher** [comment]
@AlexShaGH thanks for the details — you’re right: for train/val `imgsz` must be an integer; only predict/export accept H×W tuples.

For your 720×1280 frames:
- Train/val (single GPU for `rect=True`): keep the long side at 1280 so images pass through unchanged except for minimal stride padding.
```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt")  # same guidance applies to YOLO11 models
model.train(data="your.yaml", imgsz=1280, rect=True, device=0)
model.val(imgsz=1280, rect=True, device=0)
```
- Inference (fixed size, FP16 on GPU):
```python
model.predict(source="infer_dir", imgsz=(720, 1280), half=True, device=0)
```
- Export for edge (fixed input, FP16):
```python
model.export(format="engine", imgsz=(720, 1280), half=True)  # or format="onnx"
```

`rect=True` resizes by long side and applies minimal letterbox padding; see the rectangular resize in [BaseDataset.load_image](https://docs.ultralytics.com/reference/data/base/#ultralytics.data.base.BaseDataset.load_image) and padding details in the [LetterBox transform](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.LetterBox). Predict/export tuple support and FP16 options are in the [Predict settings](https://docs.ultralytics.com/usage/cfg/#predict-settings) and [Export settings](https://docs.ultralytics.com/usage/cfg/#export-settings).

**Comment by RajS999** [comment]
Hi @glenn-jocher , firstly, thanks for all the guidance you are providing. I have also went through your comment on another thread [here](https://github.com/ultralytics/ultralytics/issues/1037#issuecomment-1944221627). Also, I got confused, in [this comment](https://github.com/ultralytics/ultralytics/issues/9022#issuecomment-2003931837) you suggested passing tuple to `imgsz` of train(), while in [this comment](https://github.com/ultralytics/ultralytics/issues/9022#issuecomment-3305489228), you suggested passing single integer. 

I have similar doubts. Some details before my doubts:

**Some details**

1. We will be doing inference on HD and FHD stream. Note both are of 9:16 aspect ratio.
2. The training set contains images from various sources and hence they are of non-standard square size and also standard rectangular sizes (FHD, 2K, 4K and 8K)
 
I have following doubts and guesses. 

1. How should I handle different resolutions in training dataset?
    My Guess: I should crop (say with SAHI) the dataset frames to HD and FHD resolutions (doing padding wherever possible) manually before starting training.

2. Should I specify tuple or integer for `imgsz` parameter of `train()`, `export()` and `predict()`?
    My Guess: For all `train()`, `export()` and `predict()`, I should match the exact value for `imgsz` to squeeze max accuracy. Also, `imgsz` should have same (or closer) aspect ratio as the input stream, that is 9:16. So, I was guessing how about keeping `imgsz=(576, 1024)` which is exactly 9:16 and 576 is also divisible by 32 (assuming our training and inference infrastructure allows resulting latency)

3. What should be the value of `rect` parameter?
    My Guess: With manual pre-cropping, we should keep `rect=False`.

Is there any other combination of parameter values or some other approach which will let us squeeze max accuracy? Will following yield better result than above guesses:
1. `imgsz=1024`, `rect=True` and `multi_scale=True` for `train()` 
2. `imgsz=(576, 1024)` for `export()` and `predict()`?
3. No pre-cropping of HD, FHD and 2K images to HD or FHD resolutions; while pre-cropping 4K and 8K images to HD and FHD resolutions.

Can you guess any other combination of param values / approach that can yield better accuracy?

PS: We are planning to train yolo 26.

**Comment by glenn-jocher** [comment]
You’re right to call out the inconsistency — for current Ultralytics YOLO, `train()` should use a single integer `imgsz`, while `predict()` and `export()` can use either an integer or an `(h, w)` tuple; for your 9:16 deployment, a solid starting point is `train(imgsz=1024, rect=True)` and `predict()/export(imgsz=(576, 1024), rect=False)` if you need an exact fixed runtime shape, since `rect=True` uses minimum-rectangle padding and can produce tensors smaller than the tuple target. I would not manually pre-crop all HD/FHD data just to force shape consistency; keep mixed-aspect training images, and only tile/crop very large 4K/8K frames if small objects become too small after resize. This issue is closed/stale, so please open a new issue or discussion with a small sample set for visibility if you want help tuning the final setup, and see [fixed shape vs minimum rectangle](https://docs.ultralytics.com/modes/predict#fixed-shape-vs-minimum-rectangle-rect) and [image size rules](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_imgsz).
